raphtory/src/python/graph/algorithm_result.rs

```diff
@@ -0,0 +1,274 @@
+            ///
+            /// # Returns
+            ///
+            /// A sorted vector of tuples containing keys of type `H` and values of type `Y`.
+            #[pyo3(signature = (reverse=true))]
+            fn sort_by_key(&self, reverse: bool) -> Vec<($rustKey, $rustValue)> {
+                self.0.sort_by_key(reverse)
+            }
+
+            /// Retrieves the top-k elements from the `AlgorithmResult` based on its values.
+            ///
+            /// # Arguments
+            ///
+            /// * `k`: The number of elements to retrieve.
+            /// * `percentage`: If `true`, the `k` parameter is treated as a percentage of total elements.
+            /// * `reverse`: If `true`, retrieves the elements in descending order; otherwise, in ascending order.
+            ///
+            /// # Returns
+            ///
+            /// An `Option` containing a vector of tuples with keys of type `H` and values of type `Y`.
+            /// If `percentage` is `true`, the returned vector contains the top `k` percentage of elements.
+            /// If `percentage` is `false`, the returned vector contains the top `k` elements.
+            /// Returns `None` if the result is empty or if `k` is 0.
+            #[pyo3(signature = (k, percentage=false, reverse=true))]
+            fn top_k(&self, k: usize, percentage: bool, reverse: bool) -> Option<Vec<($rustKey, $rustValue)>> {
+                self.0.top_k(k, percentage, reverse)
+            }
+        }
+        py_algorithm_result_base!($name, $rustKey, $rustValue);
+    };
+}
+
+macro_rules! py_algorithm_result_ord_hash_eq {
+    ($name:ident, $rustKey:ty, $rustValue:ty) => {
+        #[pymethods]
+        impl $name {
+            /// Groups the `AlgorithmResult` by its values.
+            ///
+            /// # Returns
+            ///
+            /// A `HashMap` where keys are unique values from the `AlgorithmResult` and values are vectors
+            /// containing keys of type `H` that share the same value.
+            fn group_by(&self) -> std::collections::HashMap<$rustValue, Vec<$rustKey>> {
+                self.0.group_by()
+            }
+        }
+        py_algorithm_result_partial_ord!($name, $rustKey, $rustValue);
+    };
+}
+
+py_algorithm_result!(AlgorithmResultStrU64, String, u64);
+py_algorithm_result_ord_hash_eq!(AlgorithmResultStrU64, String, u64);
+
+py_algorithm_result!(AlgorithmResultStrTupleF32F32, String, (f32, f32));
+py_algorithm_result_partial_ord!(AlgorithmResultStrTupleF32F32, String, (f32, f32));
+
+py_algorithm_result!(AlgorithmResultStrVecI64Str, String, Vec<(i64, String)>);
+py_algorithm_result_ord_hash_eq!(AlgorithmResultStrVecI64Str, String, Vec<(i64, String)>);
+
+py_algorithm_result!(AlgorithmResultU64VecUsize, u64, Vec<usize>);
+py_algorithm_result_ord_hash_eq!(AlgorithmResultU64VecUsize, u64, Vec<usize>);
+
+py_algorithm_result!(AlgorithmResultStrF64, String, OrderedFloat<f64>);
+
+#[pymethods]
+impl AlgorithmResultStrF64 {
```

Comment:
Why not make a separate macro for this where the key can be anything

Reply:
because we cant have generic types in pyo3